### DISCUSSION TITLE
many code warnings for WW3

### DISCUSSION THREAD

**Discussion by edwardhartnett** [NOAA-EMC/WW3#1166] [General]
When I compile WW3 I get many many warnings. For example:

```
cd /home/ed/WW3_2/b/model/src && /usr/bin/gfortran -DENDIANNESS="'big_endian'" -DW3_BS0 -DW3_BT0 -DW3_CRT1 -DW3_CRX1 -DW3_DB0 -DW3_FLX2 -DW3_IC0 -DW3_IS0 -DW3_LN0 -DW3_NL0 -DW3_NOGRB -DW3_O0 -DW3_O1 -DW3_O10 -DW3_O11 -DW3_O2 -DW3_O3 -DW3_O4 -DW3_O5 -DW3_O6 -DW3_O7 -DW3_PR1 -DW3_REF0 -DW3_SHRD -DW3_ST0 -DW3_TR0 -DW3_WNT1 -DW3_WNX1 -I/home/ed/WW3_2/b/model/src/mod -I/home/ed/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/netcdf-fortran-4.5.3-grbhxyvbm2l65zjfeo7peje4zyssf5dr/include -I/home/ed/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/netcdf-c-4.7.4-cw6c5ykgvhvbtju3644xhdxk4prmxwbi/include -I/home/ed/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/mpich-3.4.2-cc4liuefusfudvvxttfbdryr7qfnu4gr/include -I/home/ed/spack/var/spack/environments/ww3-gnu/.spack-env/view/include -g -fallow-argument-mismatch -g -fno-second-underscore -ffree-line-length-none -Wall -fcheck=all -ffpe-trap=invalid,zero,overflow -frecursive -fbacktrace -c /home/ed/WW3_2/model/src/ww3_prnc.F90 -o CMakeFiles/ww3_prnc.dir/ww3_prnc.F90.o
/home/ed/WW3_2/model/src/ww3_prnc.F90:2317:4:

 2317 | 2930 FORMAT ( '          Field corrected for energy conservation.')
      |    1
Warning: Label 2930 at (1) defined but not used [-Wunused-label]
/home/ed/WW3_2/model/src/ww3_prnc.F90:2435:4:

 2435 | 1039 FORMAT (/' *** WAVEWATCH III WARNING IN W3PRNC : '/       &
      |    1
Warning: Label 1039 at (1) defined but not used [-Wunused-label]
/home/ed/WW3_2/model/src/ww3_prnc.F90:925:14:

  925 |       RW(1) = FACTOR*XGRD(1,1) ; RW(2) = FACTOR*XGRD(NY,NX)
      |              1
Warning: Possible change of value in conversion from REAL(8) to REAL(4) at (1) [-Wconversion]
/home/ed/WW3_2/model/src/ww3_prnc.F90:925:41:

  925 |       RW(1) = FACTOR*XGRD(1,1) ; RW(2) = FACTOR*XGRD(NY,NX)
      |                                         1
Warning: Possible change of value in conversion from REAL(8) to REAL(4) at (1) [-Wconversion]
/home/ed/WW3_2/model/src/ww3_prnc.F90:926:14:

  926 |       RW(3) = FACTOR*YGRD(1,1) ; RW(4) = FACTOR*YGRD(NY,NX)
      |              1
Warning: Possible change of value in conversion from REAL(8) to REAL(4) at (1) [-Wconversion]
/home/ed/WW3_2/model/src/ww3_prnc.F90:926:41:

  926 |       RW(3) = FACTOR*YGRD(1,1) ; RW(4) = FACTOR*YGRD(NY,NX)
      |                                         1
Warning: Possible change of value in conversion from REAL(8) to REAL(4) at (1) [-Wconversion]
/home/ed/WW3_2/model/src/ww3_prnc.F90:1055:14:

 1055 |           X = XGRD(1,IX)
      |              1
Warning: Possible change of value in conversion from REAL(8) to REAL(4) at (1) [-Wconversion]
/home/ed/WW3_2/model/src/ww3_prnc.F90:1056:14:

 1056 |           Y = YGRD(1,IX)
      |              1
Warning: Possible change of value in conversion from REAL(8) to REAL(4) at (1) [-Wconversion]
/home/ed/WW3_2/model/src/ww3_prnc.F90:1611:32:

 1611 |             TIDE_SECS(TIDE_NTI)=(ALLTIMES(I)-TIDE_DAYS(TIDE_NTI))*86400
      |                                1
Warning: Possible change of value in conversion from REAL(8) to INTEGER(4) at (1) [-Wconversion]
/home/ed/WW3_2/model/src/ww3_prnc.F90:1622:20:

 1622 |           TIDE_LAT= YGRD(IY,IX)
      |                    1
Warning: Possible change of value in conversion from REAL(8) to REAL(4) at (1) [-Wconversion]
/home/ed/WW3_2/model/src/ww3_prnc.F90:332:30:

  332 |   INTEGER                 :: K, L, TIDEFLAG,                      &
      |                              1
Warning: Unused variable ‘k’ declared at (1) [-Wunused-variable]
/home/ed/WW3_2/model/src/ww3_prnc.F90:332:33:

  332 |   INTEGER                 :: K, L, TIDEFLAG,                      &
      |                                 1
Warning: Unused variable ‘l’ declared at (1) [-Wunused-variable]
[ 96%] Linking Fortran executable ../../bin/ww3_prnc

```

Many of these warnings would be very easy to clear.

Ignoring code warnings is extremely counter-productive. Some of the 100+ warnings may indicate real bugs, but that cannot be seen because of all of the simple warnings that have not been cleared up. So we probably have bugs in the code, and the compiler is telling us about them, but we are ignoring that. What a waste, because finding a bug is the most expensive part of maintenance, and the most expensive part of programming. So when the compiler is telling us about problems, and we just ignore it, it's really wasteful.

NCO has a rule, only sporadically enforced, that there should be no compile warnings in operational code. It's a great rule and they should enforce it strictly, IMO. We have turned on warnings checks in the CI of the NCEPLIBS, so that no code which causes warnings (for GNU, with -Wall) can be merged. Clearing up all the warnings in the code left by careless programmers over the years didn't take long.

Allowing warnings in your code is like allowing litter on your street. Not only does it look bad, but it also gives permission for everyone else to litter on your street. (See the [broken windows theory](https://en.wikipedia.org/wiki/Broken_windows_theory#:~:text=In%20criminology%2C%20the%20broken%20windows,and%20disorder%2C%20including%20serious%20crimes.).) In WW3, you are giving permission for every programmer to check in code full of warnings.

Wouldn't the software benefit, programmers be more productive, and the science possibly more accurate if the code compiled cleanly, without warning?

**Comment by ukmo-ccbunney** [comment]
@edwardhartnett I think a lot of these warnings (certainly those regarding unused variables, labels, etc) are due to the prodigious use of `#ifdef` directives to handle all the different compile "switches" that WW3 has (i.e. the switches such as ST3, ST4, PR2, PR3, etc).
When certain switches are not used, the code blocks are removed by the pre-processor, but more often than not the variable declarations are not protected by the same `#ifdef` directives.

I think remedying the issue is a bit of a double-edge sword: adding lots more `#ifdef` declarations will quieten the compiler warnings, but makes the code harder to read. Some modules are affected by this more than others. W3SRCE is quite bad, but I am trying to remedy this with my refactor work (W3SRCE also has lots and lots of #ifdef directives!)

That's not to say that I don't agree with your "no compiler warnings" ethos - I think that is definitely something we should strive to achieve! It's just that this does come with a downside!

Note: Obviously, some of the warnings can (and should) be fixed, such as the precision conversion ones...

**Reply by edwardhartnett** [reply]
I would prefer more ifdefs and no warnings.

There is an overuse of ifdefs in the code. I understand this is a historical legacy. However, it has consequences, and this is one of them. The answer, IMO, is not to build tolerance for the warnings, but to use ifdef where needed, and work towards a refactoring where if/endif is used instead in many cases.
